<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from model.urdf.xacro               | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="gazebo_ros_four_wheel_steering_test_vehicle">
  <link name="chassis">
    <inertial>
      <mass value="100.0"/>
      <origin xyz="-0.05 0 -0.1"/>
      <inertia ixx="8.33333" ixy="0.0" ixz="0.0" iyy="36.3333188" iyz="0.0" izz="38.666651200000004"/>
    </inertial>
    <visual>
      <geometry>
        <box size="2 0.8 0.6"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <box size="2 0.8 0.6"/>
      </geometry>
    </collision>
  </link>
  <link name="front_left_axle">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://metron_p48_description/media/models/wheels/motor.dae"/>
      </geometry>
      <material name="gray">
        <color rgba="0.4 0.4 0.4 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10.0"/>
      <origin xyz="0 -0.1 0"/>
      <inertia ixx="0.04000000000000001" ixy="0.0" ixz="0.0" iyy="0.04000000000000001" iyz="0.0" izz="0.04000000000000001"/>
    </inertial>
  </link>
  <joint name="front_left_kingpin" type="revolute">
    <limit effort="1000" lower="-0.6109" upper="0.6109" velocity="1.4"/>
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="front_left_axle"/>
    <origin xyz="0.8 0 0"/>
    <dynamics damping="1.0" friction="0"/>
    <gazebo>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </gazebo>
  </joint>
  <link name="front_right_axle">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://metron_p48_description/media/models/wheels/motor.dae"/>
      </geometry>
      <material name="gray">
        <color rgba="0.4 0.4 0.4 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10.0"/>
      <origin xyz="0 0.1 0"/>
      <inertia ixx="0.04000000000000001" ixy="0.0" ixz="0.0" iyy="0.04000000000000001" iyz="0.0" izz="0.04000000000000001"/>
    </inertial>
  </link>
  <joint name="front_right_kingpin" type="revolute">
    <limit effort="1000" lower="-0.6109" upper="0.6109" velocity="1.4"/>
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="front_right_axle"/>
    <origin xyz="0.8 0 0"/>
    <dynamics damping="1.0" friction="0"/>
    <gazebo>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </gazebo>
  </joint>
  <!-- SDF-only steering system -->
  <gazebo>
    <link name="front_steering_gear">
      <pose>
          1.28 0 0 0 0 0
        </pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0016666660000000002</ixx>
          <iyy>0.0016666660000000002</iyy>
          <izz>0.0016666660000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="front_visual">
        <geometry>
          <cylinder>
            <radius>0.1</radius>
            <length>0.1</length>
          </cylinder>
        </geometry>
      </visual>
      <collision name="front_collision">
        <geometry>
          <cylinder>
            <radius>0.1</radius>
            <length>0.1</length>
          </cylinder>
        </geometry>
      </collision>
    </link>
    <joint name="front_steering" type="revolute">
      <parent>chassis</parent>
      <child>front_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
        <limit>
          <lower>-1.5</lower>
          <upper>1.5</upper>
          <effort>2000</effort>
          <velocity>1.24</velocity>
        </limit>
        <dynamics>
          <damping>1</damping>
          <friction>2</friction>
        </dynamics>
      </axis>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </joint>
    <link name="front_left_steering_gear">
      <pose>1.28 0 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="front_left_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="front_left_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <link name="front_left_axle_rod_connector">
      <pose>1.18 0.075 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="front_left_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="front_left_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <joint name="front_left_steering_connector" type="revolute">
      <parent>front_left_steering_gear</parent>
      <child>front_left_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="front_left_steering_rod" type="revolute">
      <parent>front_left_steering_gear</parent>
      <child>front_left_axle_rod_connector</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="front_left_axle_connection" type="fixed">
      <parent>front_left_axle_rod_connector</parent>
      <child>front_left_axle</child>
    </joint>
    <link name="front_right_steering_gear">
      <pose>1.28 0 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="front_right_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="front_right_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <link name="front_right_axle_rod_connector">
      <pose>1.18 -0.075 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="front_right_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="front_right_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <joint name="front_right_steering_connector" type="revolute">
      <parent>front_right_steering_gear</parent>
      <child>front_right_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="front_right_steering_rod" type="revolute">
      <parent>front_right_steering_gear</parent>
      <child>front_right_axle_rod_connector</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="front_right_axle_connection" type="fixed">
      <parent>front_right_axle_rod_connector</parent>
      <child>front_right_axle</child>
    </joint>
  </gazebo>
  <link name="rear_left_axle">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://metron_p48_description/media/models/wheels/motor.dae"/>
      </geometry>
      <material name="gray">
        <color rgba="0.4 0.4 0.4 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10.0"/>
      <origin xyz="0 -0.1 0"/>
      <inertia ixx="0.04000000000000001" ixy="0.0" ixz="0.0" iyy="0.04000000000000001" iyz="0.0" izz="0.04000000000000001"/>
    </inertial>
  </link>
  <joint name="rear_left_kingpin" type="revolute">
    <limit effort="1000" lower="-0.6109" upper="0.6109" velocity="1.4"/>
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="rear_left_axle"/>
    <origin xyz="-0.8 0 0"/>
    <dynamics damping="1.0" friction="0"/>
    <gazebo>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </gazebo>
  </joint>
  <link name="rear_right_axle">
    <visual>
      <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://metron_p48_description/media/models/wheels/motor.dae"/>
      </geometry>
      <material name="gray">
        <color rgba="0.4 0.4 0.4 1.0"/>
      </material>
    </visual>
    <collision>
      <geometry>
        <sphere radius="0.2"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="10.0"/>
      <origin xyz="0 0.1 0"/>
      <inertia ixx="0.04000000000000001" ixy="0.0" ixz="0.0" iyy="0.04000000000000001" iyz="0.0" izz="0.04000000000000001"/>
    </inertial>
  </link>
  <joint name="rear_right_kingpin" type="revolute">
    <limit effort="1000" lower="-0.6109" upper="0.6109" velocity="1.4"/>
    <axis xyz="0 0 1"/>
    <parent link="chassis"/>
    <child link="rear_right_axle"/>
    <origin xyz="-0.8 0 0"/>
    <dynamics damping="1.0" friction="0"/>
    <gazebo>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </gazebo>
  </joint>
  <!-- SDF-only steering system -->
  <gazebo>
    <link name="rear_steering_gear">
      <pose>
          1.28 0 0 0 0 0
        </pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0016666660000000002</ixx>
          <iyy>0.0016666660000000002</iyy>
          <izz>0.0016666660000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="rear_visual">
        <geometry>
          <cylinder>
            <radius>0.1</radius>
            <length>0.1</length>
          </cylinder>
        </geometry>
      </visual>
      <collision name="rear_collision">
        <geometry>
          <cylinder>
            <radius>0.1</radius>
            <length>0.1</length>
          </cylinder>
        </geometry>
      </collision>
    </link>
    <joint name="rear_steering" type="revolute">
      <parent>chassis</parent>
      <child>rear_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
        <limit>
          <lower>-1.5</lower>
          <upper>1.5</upper>
          <effort>2000</effort>
          <velocity>1.24</velocity>
        </limit>
        <dynamics>
          <damping>1</damping>
          <friction>2</friction>
        </dynamics>
      </axis>
      <physics>
        <ode>
          <cfm_damping>1</cfm_damping>
        </ode>
      </physics>
    </joint>
    <link name="rear_left_steering_gear">
      <pose>-1.28 0 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="rear_left_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="rear_left_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <link name="rear_left_axle_rod_connector">
      <pose>-1.18 -0.075 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="rear_left_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="rear_left_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <joint name="rear_left_steering_connector" type="revolute">
      <parent>rear_left_steering_gear</parent>
      <child>rear_left_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="rear_left_steering_rod" type="revolute">
      <parent>rear_left_steering_gear</parent>
      <child>rear_left_axle_rod_connector</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="rear_left_axle_connection" type="fixed">
      <parent>rear_left_axle_rod_connector</parent>
      <child>rear_left_axle</child>
    </joint>
    <link name="rear_right_steering_gear">
      <pose>-1.28 0 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="rear_right_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="rear_right_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <link name="rear_right_axle_rod_connector">
      <pose>-1.18 0.075 0 0 0 0</pose>
      <inertial>
        <mass>1</mass>
        <inertia>
          <ixx>0.0010000000000000002</ixx>
          <iyy>0.0010000000000000002</iyy>
          <izz>0.0010000000000000002</izz>
          <ixy>0.0</ixy>
          <ixz>0.0</ixz>
          <iyz>0.0</iyz>
        </inertia>
      </inertial>
      <visual name="rear_right_visual">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </visual>
      <collision name="rear_right_collision">
        <geometry>
          <sphere>
            <radius>0.05</radius>
          </sphere>
        </geometry>
      </collision>
    </link>
    <joint name="rear_right_steering_connector" type="revolute">
      <parent>rear_right_steering_gear</parent>
      <child>rear_right_steering_gear</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="rear_right_steering_rod" type="revolute">
      <parent>rear_right_steering_gear</parent>
      <child>rear_right_axle_rod_connector</child>
      <axis>
        <xyz>0 0 1</xyz>
      </axis>
    </joint>
    <joint name="rear_right_axle_connection" type="fixed">
      <parent>rear_right_axle_rod_connector</parent>
      <child>rear_right_axle</child>
    </joint>
  </gazebo>
  <link name="front_left_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="30.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.79657088137152" ixy="0.0" ixz="0.0" iyy="0.79657088137152" iyz="0.0" izz="1.1726424"/>
    </inertial>
  </link>
  <joint name="front_left_wheel" type="continuous">
    <parent link="front_left_axle"/>
    <child link="front_left_wheel_link"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 0.156911 0"/>
    <limit effort="755" velocity="6.358"/>
    <axis xyz="0 0 -1"/>
    <dynamics friction="12"/>
  </joint>
  <link name="front_right_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="30.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.79657088137152" ixy="0.0" ixz="0.0" iyy="0.79657088137152" iyz="0.0" izz="1.1726424"/>
    </inertial>
  </link>
  <joint name="front_right_wheel" type="continuous">
    <parent link="front_right_axle"/>
    <child link="front_right_wheel_link"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 -0.156911 0"/>
    <limit effort="755" velocity="6.358"/>
    <axis xyz="0 0 -1"/>
    <dynamics friction="12"/>
  </joint>
  <link name="rear_left_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="30.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.79657088137152" ixy="0.0" ixz="0.0" iyy="0.79657088137152" iyz="0.0" izz="1.1726424"/>
    </inertial>
  </link>
  <joint name="rear_left_wheel" type="continuous">
    <parent link="rear_left_axle"/>
    <child link="rear_left_wheel_link"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 0.156911 0"/>
    <limit effort="755" velocity="6.358"/>
    <axis xyz="0 0 -1"/>
    <dynamics friction="12"/>
  </joint>
  <link name="rear_right_wheel_link">
    <visual>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.29" radius="0.3495"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="30.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.79657088137152" ixy="0.0" ixz="0.0" iyy="0.79657088137152" iyz="0.0" izz="1.1726424"/>
    </inertial>
  </link>
  <joint name="rear_right_wheel" type="continuous">
    <parent link="rear_right_axle"/>
    <child link="rear_right_wheel_link"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 -0.156911 0"/>
    <limit effort="755" velocity="6.358"/>
    <axis xyz="0 0 -1"/>
    <dynamics friction="12"/>
  </joint>
  <gazebo>
    <plugin filename="libgazebo_ros_four_wheel_steering.so" name="gazebo_ros_4ws">
      <ros>
        <namespace>test</namespace>
      </ros>
      <front_right_motor_pid_gain>400 0 1</front_right_motor_pid_gain>
      <front_left_motor_pid_gain>400 0 1</front_left_motor_pid_gain>
      <rear_right_motor_pid_gain>400 0 1</rear_right_motor_pid_gain>
      <rear_left_motor_pid_gain>400 0 1</rear_left_motor_pid_gain>
      <front_steering_pid_gain>2000 0 2</front_steering_pid_gain>
      <rear_steering_pid_gain>2000 0 2</rear_steering_pid_gain>
    </plugin>
  </gazebo>
</robot>
